File: leetcode/1-100/59-spiral-matrix-ii/bench/spiral_matrix_ii.c
// Benchmark workload — Spiral Matrix II (LeetCode #59).
// C mirror of bench/spiral_matrix_ii.kara. Same M=9 rotated sizes (n=12..20),
// K=180k, boundary-shrinking generator, position-weighted checksum, and sink.
// The grid is int64_t** (a row-pointer array + one malloc per row = n+1
// allocations), matching the kara/rust Vec<Vec<i64>> nested-allocation shape
// — apples-to-apples on allocation, not a single flat buffer.

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

static int64_t **generate_matrix(int64_t n) {
    int64_t **grid = (int64_t **)malloc(sizeof(int64_t *) * (size_t)n);
    for (int64_t r = 0; r < n; r++) {
        grid[r] = (int64_t *)calloc((size_t)n, sizeof(int64_t));
    }

    int64_t top = 0, bottom = n - 1, left = 0, right = n - 1, v = 1;
    while (top <= bottom && left <= right) {
        for (int64_t c = left; c <= right; c++) {
            grid[top][c] = v++;
        }
        top++;

        for (int64_t r2 = top; r2 <= bottom; r2++) {
            grid[r2][right] = v++;
        }
        right--;

        if (top <= bottom) {
            for (int64_t c2 = right; c2 >= left; c2--) {
                grid[bottom][c2] = v++;
            }
            bottom--;
        }

        if (left <= right) {
            for (int64_t r3 = bottom; r3 >= top; r3--) {
                grid[r3][left] = v++;
            }
            left++;
        }
    }
    return grid;
}

static int64_t checksum(int64_t **grid, int64_t n) {
    int64_t s = 0;
    for (int64_t i = 0; i < n; i++) {
        for (int64_t j = 0; j < n; j++) {
            s += grid[i][j] * (i * n + j + 1);
        }
    }
    return s;
}

static void free_matrix(int64_t **grid, int64_t n) {
    for (int64_t r = 0; r < n; r++) free(grid[r]);
    free(grid);
}

int main(void) {
    const int64_t m_sizes = 9;
    const int64_t k_iters = 180000;

    int64_t total = 0;
    for (int64_t k = 0; k < k_iters; k++) {
        int64_t n = 12 + (k % m_sizes);
        int64_t **g = generate_matrix(n);
        total += checksum(g, n);
        free_matrix(g, n);
    }
    printf("%lld\n", (long long)total);
    return 0;
}

File: leetcode/1-100/59-spiral-matrix-ii/bench/spiral_matrix_ii.py
"""Benchmark workload — Spiral Matrix II (LeetCode #59).

Python mirror of bench/spiral_matrix_ii.kara. Same M=9 rotated sizes
(n=12..20), boundary-shrinking generator over a list-of-lists, position-
weighted checksum, and sink. CPython is multi-second per sample at K=180k, so
this mirror runs K=18_000 (1/10th); K is a multiple of 90 so each of the 9
sizes is hit exactly 2_000 times and the sink is exactly 1/10th of the
compiled mirrors' sink — the README quotes the projected ratio.
"""


def generate_matrix(n: int) -> list[list[int]]:
    grid = [[0] * n for _ in range(n)]
    top, bottom, left, right = 0, n - 1, 0, n - 1
    v = 1

    while top <= bottom and left <= right:
        for c in range(left, right + 1):
            grid[top][c] = v
            v += 1
        top += 1

        for r in range(top, bottom + 1):
            grid[r][right] = v
            v += 1
        right -= 1

        if top <= bottom:
            for c in range(right, left - 1, -1):
                grid[bottom][c] = v
                v += 1
            bottom -= 1

        if left <= right:
            for r in range(bottom, top - 1, -1):
                grid[r][left] = v
                v += 1
            left += 1

    return grid


def checksum(grid: list[list[int]], n: int) -> int:
    s = 0
    for i in range(n):
        for j in range(n):
            s += grid[i][j] * (i * n + j + 1)
    return s


def main() -> None:
    m_sizes = 9
    k_iters = 18_000

    total = 0
    for k in range(k_iters):
        n = 12 + (k % m_sizes)
        g = generate_matrix(n)
        total += checksum(g, n)
    print(total)


if __name__ == "__main__":
    main()
